File: 1/2020.09.21/42.Trapping Rain Water/java/solution.java
class Solution {
    public int trap(int[] height) {
        int sum = 0;
        int left = 0, right = height.length - 1;
        int max_left = 0, max_right = 0;
        while (left < right) {
            if (height[left] < height[right]) {
                if (height[left] >= max_left) {
                    max_left = height[left];
                } else {
                    sum += max_left - height[left];
                }
                left++;
            } else {
                if (height[right] >= max_right) {
                    max_right = height[right];
                } else {
                    sum += max_right - height[right];
                }
                right--;
            }
        }
        return sum;
    }
}

File: 1/2020.09.21/42.Trapping Rain Water/python/solution.py
class Solution:
    def trap(self, height: List[int]) -> int:
        # double pointer
        sum = max_left = max_right = 0
        left, right = 0, len(height) - 1
        while left < right:
            if height[left] < height[right]:
                if height[left] >= max_left:
                    max_left = height[left]
                else:
                    sum += max_left - height[left]
                left += 1
            else:
                if height[right] >= max_right:
                    max_right = height[right]
                else:
                    sum += max_right - height[right]
                right -= 1
        return sum
